### ISSUE TITLE
Add support to conda lock files

### ISSUE THREAD

**Issue by remidebette** [aquasecurity/trivy#1856]
Hi,

Additionally to #492, would it be possible to add support to conda [explicit files](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#building-identical-conda-environments) (as obtained by `conda list --explicit`)

These files can also be generated by the package [conda-lock](https://github.com/conda-incubator/conda-lock), so in the end they cover pretty well conda in general.

Also, I understand from [here](https://github.com/conda-forge/conda-forge.github.io/issues/1486) that conda-forge support can be tricky, do you think this can happen with trivy?

**Comment by knqyf263** [comment]
Let's see how many users need the conda support. Also, we welcome contribution from anyone.

**Comment by remidebette** [comment]
Still needed

**Comment by remidebette** [comment]
Still needed

**Comment by maitre-matt** [comment]
I am new to Trivy (or Go for that matter...), but with a few code pointers I can likely take a stab at a contribution. For pip, it looks like Trivy builds an inventory of `.dist-info/META-DATA` files. Digging into conda environments on my machine, it looks like the conda equivalent is JSON files like `<conda-root>/envs/<env>/conda-meta/<package>.json`. The [conda doc](https://conda.io/projects/conda/en/latest/user-guide/concepts/environments.html) is pretty vague about those, but assuming that this is a generic pattern and not just on my machine, it should be easy to parse.

The main interest for that feature on my side is generating SBOMs that include conda packages since conda is often used to manage Machine Learning code and tracking which versions of Python, etc. are in used is best done by querying conda.

Beyond package inventory, the conda docs mention an [API](https://server-docs.anaconda.com/en/latest/admin/cve.html) to get CVEs for conda packages so vulnerability reporting could likely be added too if not already in place.

**Comment by maitre-matt** [comment]
My understanding of the work so far:
- Add a parser along the lines of https://github.com/aquasecurity/go-dep-parser/tree/main/pkg/python/packaging
- Add an analyzer https://github.com/aquasecurity/trivy/tree/57ed5774590434acc8becfb131338f5889cab749/pkg/fanal/analyzer/language/python/packaging
- Maybe add a detector (likely out of scope) https://github.com/aquasecurity/trivy/tree/827cea3245081923df9b0e06e5e4f33fbe7302f3/pkg/detector/library
- Update doc https://github.com/aquasecurity/trivy/blob/4839075c284f5ede6d8ed897a14e1d27388a3c3a/docs/docs/vulnerability/detection/language.md

**Comment by maitre-matt** [comment]
@knqyf263 : would someone be able to review PR https://github.com/aquasecurity/go-dep-parser/pull/151?

**Comment by knqyf263** [comment]
Sure

**Comment by maitre-matt** [comment]
@knqyf263: would you be able to do the second review on PR https://github.com/aquasecurity/go-dep-parser/pull/151? The first review is done.

**Comment by knqyf263** [comment]
Hi @maitre-matt, I'm on it now. Just to clarify, you are trying to add support for conda packages, not conda lock files, right?

>Beyond package inventory, the conda docs mention an [API](https://server-docs.anaconda.com/en/latest/admin/cve.html) to get CVEs for conda packages so vulnerability reporting could likely be added too if not already in place.

Looks like Anaconda has their own security advisories. It means we may be unable to use security advisories for pip, such as [GHSA](https://github.com/advisories?query=type%3Areviewed+ecosystem%3Apip). This is because Anaconda may have different versioning from upstream and add some changes to the source code. If we use GHSA for conda packages by mistake, it leads to false detection.

>In its default configuration, conda can install and manage the thousand packages at [repo.anaconda.com](https://repo.anaconda.com/) that are built, reviewed and maintained by Anaconda®.

They seem to build packages themselves. We need to clarify which security advisories should be used for conda packages.

> The main interest for that feature on my side is generating SBOMs that include conda packages

If you just want to include conda packages in SBOM, we can go for it. In that case, conda packages will not be used for vulnerability detection.

**Comment by maitre-matt** [comment]
@knqyf263 : the change indeed detects conda packages and not conda lock files. If this creates confusion, I can open a separate issue on GitHub. In pip world this would be equivalent to detecting `dist-info` instead of `requirements.txt`.

GHSA currently does not track vulnerabilities on conda packages and as you mentioned trying to map conda packages to pip packages will lead to errors. The two should be treated as separate in terms of vulnerabilities. Conda and pip form a Venn diagram: some conda packages do not exist in pip (ex: Python itself) and vice versa.

For the time being I'd like to focus on SBOM generation. If I'm able to ramp up on CVE handling in vuln-list-update/trivy-db I may be able to tackle the more generic vulnerability detection.

**Comment by knqyf263** [comment]
> If this creates confusion, I can open a separate issue on GitHub.

Yes, it sounds better. This issue is for conda lock files. We treat conda packages and conda lock files separately, like `Pipfile.lock` and `dist-info`.

> For the time being I'd like to focus on SBOM generation.

Got it. Could you create a new issue for conda packages and SBOM?

We also need to update the following page.
https://aquasecurity.github.io/trivy/dev/docs/sbom/supported/

**Comment by maitre-matt** [comment]
@knqyf263 : here is the new issue: #3321

**Comment by samuraixwandering** [comment]
I'd like to take this.

Remaining work is **conda lock / explicit pin files** only (not `conda-meta` / `environment.yml`, already supported). First cut: **`conda list --explicit`**-style files for SBOM inventory, same as other conda (no new vuln DB). Docs + tests in the PR.

Please say if that's already claimed.
